We implemented the division-based Euklid algorithm

In [1]:
function OwnGcd(a,b)
    while b ≠ 0
        t = b
        b = a %b
        a = t
    end
    return abs(a)
end

OwnGcd (generic function with 1 method)

To test the code we generate random integers and compare our fdunction to the build in one.

In [2]:
for _ in 1:10
    a = rand(Int16)
    b = rand(Int16)
    @assert OwnGcd(a,b) == gcd(a,b)
end

In [3]:
function findPeriod(a,N)
    x = 1
    result = a % N
    while result != 1
        result = (result * a) % N
        x += 1
    end
    return x
end

findPeriod (generic function with 1 method)

In [4]:
function findPrimeFactors(N)
    while true
    a = rand(1:N-1)
    d = OwnGcd(a, N)
    if d != 1
        return d
    end
    r = findPeriod(a, N)
    if r % 2 == 1
        continue
    end
    a_r_half = powermod(a, r ÷ 2, N)
    if a_r_half == N - 1
        continue
    end
    return OwnGcd(a_r_half + 1, N), OwnGcd(a_r_half - 1, N)
end
end

findPrimeFactors (generic function with 1 method)

The extended Euclidean algorithm is basen on [Source](https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm#Computing_multiplicative_inverses_in_modular_structures)

In [5]:
function inverse(a,n)
    t = 0; r=n
    t_new = 1; r_new =a
    while r_new ≠ 0

        qt = ÷(r,r_new)
        t, t_new = (t_new, t- qt*t_new)
        r, r_new = (r_new, r- qt*r_new)

    end
    @assert !(r>1)
    return t*(1+n*(t<0))
end

inverse (generic function with 1 method)

In [6]:
ϕ(p,q) = (p-1)*(q-1)

ϕ (generic function with 1 method)

In [7]:
N = 71219839
e = 4187899

4187899

In [8]:
@show p, q = findPrimeFactors(N)
@assert p*q == N

(p, q) = findPrimeFactors(N) = (7649, 9311)


In [9]:
d =inverse(e,ϕ(p,q))

6567859

Thus the decryption Key is d = 6567859, and p = 7649, q = 9311

**2.**

To de/encode Dicts are used

In [10]:
key = Dict([(Char(97 +i  ),i) for i in 0:25])
rkey = Dict(v => k for (k,v) in key)

Dict{Int64, Char} with 26 entries:
  5  => 'f'
  16 => 'q'
  20 => 'u'
  12 => 'm'
  24 => 'y'
  8  => 'i'
  17 => 'r'
  23 => 'x'
  19 => 't'
  22 => 'w'
  0  => 'a'
  6  => 'g'
  1  => 'b'
  11 => 'l'
  9  => 'j'
  14 => 'o'
  3  => 'd'
  7  => 'h'
  25 => 'z'
  ⋮  => ⋮

In [11]:
function getEncodedMassage(string,keys)
    encodedMessage = 0
    for (i,char) in enumerate(reverse(string))
        encodedMessage += key[char]*26^(i-1)
    end
    return encodedMessage
end
function getDecodedMassage(int, keys)
    message = ""
    while int > 0
         message = keys[(int % 26)] * message
        int ÷= 26
    end
    return message
end

getDecodedMassage (generic function with 1 method)

To test the encoder I use the example of the sheet:

In [12]:
@show testEncode = getEncodedMassage("hallo",key)
getDecodedMassage(testEncode, rkey)

testEncode = getEncodedMassage("hallo", key) = 3206568


"hallo"

I can now encode the string "bwtlqv":

In [13]:
encryptedEncodedMsg = getEncodedMassage("bwtlqv",key)

22276665

I can decrypt the massage using the decryption Key d and N:

In [14]:
decryptedEncodedMsg = powermod(encryptedEncodedMsg, d,N)

203179

Finally I can decode the decrypted massage

In [15]:
getDecodedMassage(decryptedEncodedMsg,rkey)

"loop"

The secret massage is thus "loop"